rafthttp/snapshot.go

```diff
@@ -0,0 +1,188 @@
+// then let remote member send request to ask for snapshot, so at least
+// two maximal RTT is required, which is 10s.
+const snapshotWaitTimeout = 10 * time.Second
+
+type urlPickerGetter interface {
+	getURLPicker(id types.ID) *urlPicker
+}
+
+func newSnapshotBuffer(r Raft, snapst SnapshotStore) *snapshotBuffer {
+	return &snapshotBuffer{
+		r:      r,
+		snapst: snapst,
+	}
+}
+
+func newSnapshotProcessor(tr http.RoundTripper, upGetter urlPickerGetter, r Raft, snapst SnapshotStore) *snapshotProcessor {
+	return &snapshotProcessor{
+		tr:       tr,
+		upGetter: upGetter,
+		r:        r,
+		snapst:   snapst,
+	}
+}
+
+// snapshotBuffer represents a buffer of snapshot with write method.
+type snapshotBuffer struct {
+	r      Raft
+	snapst SnapshotStore
+
+	// protect fields below
+	mu sync.Mutex
+	// whether it has snapshot
+	hasSnap bool
+	// the member ID that the snapshot is sent to
+	snapTo types.ID
+	// the index of the snapshot
+	snapIndex uint64
+	// when the timer expires, it will call the function attached via AfterFunc
+	// to close the snapshot
+	snapCloseTimer *time.Timer
+}
+
+// addSnap adds snapshot into snapshot buffer.
+// The added snapshot waits to be written out through writeSnapTo func.
+// If the snapshot exceeds wait timeout, it is closed and snapshot buffer removes it.
+// At most one snapshot could be in the buffer at the same time, or it panics.
+func (s *snapshotBuffer) addSnap(to types.ID, index uint64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.hasSnap {
+		plog.Panicf("unexpected add snapshot when previous one has not finished")
+	}
+	s.hasSnap = true
+	s.snapTo = to
+	s.snapIndex = index
+	s.snapCloseTimer = time.AfterFunc(snapshotWaitTimeout, func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		if !s.hasSnap {
+			return
+		}
+		s.hasSnap = false
+		// close the snapshot in the store
+		s.snapst.WriteSnapTo(&errWriter{errors.New("cancelled")})
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out due to timeout", index, s.snapTo)
+		s.r.ReportSnapshot(uint64(to), raft.SnapshotFailure)
+	})
+}
+
+// writeSnapTo writes the snapshot with the given index into the given writer.
+func (s *snapshotBuffer) writeSnapTo(w io.Writer, index uint64) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.hasSnap {
+		return fmt.Errorf("failed to find available snapshot")
+	}
+	if s.snapIndex != index {
+		return fmt.Errorf("failed to find the snapshot with index %d", index)
+	}
+	s.hasSnap = false
+	s.snapCloseTimer.Stop()
+
+	err := s.snapst.WriteSnapTo(w)
+	if err != nil {
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out (%v)", index, s.snapTo, err)
+		s.r.ReportSnapshot(uint64(s.snapTo), raft.SnapshotFailure)
+		return err
+	}
+	plog.Infof("snapshot [index: %d, to: %s] sent out successfully", index, s.snapTo)
+	s.r.ReportSnapshot(uint64(s.snapTo), raft.SnapshotFinish)
+	return nil
+}
+
+// snapshotProcessor processes incoming raft snapshot message.
```

Comment:
// snapshotProcessor is used to process raft snapshot.
// When receiving a snapMsg, the transport layer should send the snapMsg to snapshotProcessor.
// snapshotProcessor helps to persist the snapshot and then ask raft state machine to further process 
// the snapMsg.